File: szabo_grid.m
%max allowed density is 1.39
% rho = (pi * Req^2 * N) / L^2
% delX^2 * N = L^2
% rhoMax = 1.39 when delX = 1.5*Req

function orderN=szabo_grid(rho,noise)

numberOfPoints = 100;N=numberOfPoints; % must be perfect square
v0 = 1; %self-propelling speed
tau = 1;mu = 1; %parameters
Req = 5/6; %equilibrium radius
R0 = 1; %cut-off radius

Fadh = 0.75;Frep = 30; %force parameters
% Fwall = 50;

L = Req * sqrt(pi*numberOfPoints/rho); %domain size (side length)
delX = L/sqrt(N); %grid size (length)

x = delX/2:delX:delX*(sqrt(N)-1)+delX/2; %x positions
keeperX = repmat(x,1,sqrt(N)); %sqrt(N) rows of particles parallel to x axis, listed as a 1d array; sqrt(N)*sqrt(N) = N entries
y = x; %y positions, identical distribution as for x
tempY = repmat(y,sqrt(N),1);
keeperY = reshape(tempY,1,N); %reshape sqrt(N)xsqrt(N) array into a 1d array, with N entries, each representing one particle

Nsteps=20000; %number of timeSteps
% cutoffIter=Nsteps-100; %not being used currently
%timedelta=0.05*R0/v0; %not being used currently
theta = 2*pi*rand(1,numberOfPoints) - pi; %need random direction of motion, in the range of -pi to pi
timedelta=0.005; %time stepping interval

y=zeros(Nsteps+1,3*numberOfPoints); % solution matrix, initialised with a bunch of zeros
y(1,:) = [keeperX,keeperY,theta]; % initial position
%----------------Explanation of the structure of the solution matrix    y: ---------------
%each row represents a single time instant
%first N entries in a row represent the N x-coordinates at that time instant
%next N entries are the corresponding N y-coordinates 
%last N entries are the corresponding angle that the velocity vector makes with the x axis
%-----------------------------------------------------------------------------------------

orderN = zeros(1,Nsteps); %stores order parameter at each time step
for k=1:Nsteps
	posX=y(k,1:numberOfPoints); % x position matrix
	posY=y(k,numberOfPoints+1:2*numberOfPoints); % y position matrix
	theta=y(k,2*numberOfPoints+1:end); % direction matrix
    
	drift=[v0*cos(theta) v0*sin(theta)]; 
    % first N entries store x components of velocity, next N store the corresponding y components
    
    rhs = zeros(1,2*numberOfPoints); % rhs of the position updation Diff. eqn
    % structure of rhs exactly identical to that of drift
    % first N entries store x components of force, next N store the corresponding y components
    
    for i=1:numberOfPoints %iterating over each point
        interaxn=[0,0]; %initialise force of interaction, i.e. [Fx Fy]
        for j=1:numberOfPoints % to calculate the interaction force on i due to j
            if j==i
                continue;
            end
            dx = posX(i) - posX(j);
			dy = posY(i) - posY(j);
            %periodicity conditions
            if abs(dx) > 0.5*L
				dx = dx - L*sign(dx);
            end
            if abs(dy) > 0.5*L
				dy = dy - L*sign(dy);
            end
            
            dijSq = dx^2 + dy^2; % square of distance between i and j
            dij = sqrt(dijSq);
            eij = [dx,dy]/dij; %unit vector e_{ij} joining i and j
            if dijSq<Req^2 %attractive interaction force
                interaxn = interaxn - Frep*(dij-Req)/Req * eij;
            elseif dijSq>=Req^2 && dijSq<=R0^2 %repulsive interaction force
                interaxn = interaxn - Fadh*(dij-Req)/(R0-Req) * eij;
            end
        end
    %    %for wall force (in case wall is included)
    %         Fx = 0;Fy = 0;
    %         if posX(i)<R0
    %             Fx = Fx + Fwall*exp(-2*posX(i)/R0);
    %         end
    %         if L-posX(i)<R0
    %             Fx = Fx - Fwall*exp(-2*(L-posX(i))/R0);
    %         end
    %         if posY(i)<R0
    %             Fy = Fy + Fwall*exp(-2*posY(i)/R0);
    %         end
    %         if L-posY(i)<R0
    %             Fy = Fy - Fwall*exp(-2*(L-posY(i))/R0);
    %         end
    %         interaxn = interaxn + [Fx Fy];
        interaxn=mu*interaxn; % multiply by mobility parameter
        rhs(i)=drift(i)+interaxn(1); %x components, i.e. first N entries of rhs
        rhs(numberOfPoints+i)=drift(numberOfPoints+i)+interaxn(2); % y components
    end
    s=[0,0]; %avg velocity initialise
    res = zeros(1,numberOfPoints); %rhs for the angle updation diff eqn
    for i=1:numberOfPoints
        ni=[drift(i) drift(numberOfPoints+i)]/v0; %unit vector in direction of self-propulsion of i-th particle
        vi=[rhs(i) rhs(numberOfPoints+i)]; %velocity vector of i-th particle
        normvi=sqrt(vi(1)^2+vi(2)^2);
        vi=vi/normvi; %unit vector in the direction of velocity of i-th particle
        s=s+[vi(1) vi(2)]; %calculation of avg velocity (for calculation of order parameter)
        res(i)=asin(ni(1)*vi(2)-ni(2)*vi(1)); %calculation of the rhs of i-th particel theta diff. eqn
    end
    orderN(k)=sqrt(s(1)^2+s(2)^2)/numberOfPoints; %order parameter at the k-th step
	res = (1/tau)*res; %parameter tau being factored in
    %updation of position as per Euler stepping scheme
	posX = posX + timedelta*rhs(1:numberOfPoints);
    posY = posY + timedelta*rhs(numberOfPoints+1:end);
    %periodic boundary conditions
	posX = posX - L * floor(posX/L);
	posY = posY - L * floor(posY/L);
% 	theta = theta + timedelta*res + sqrt(timedelta*(noise^2)/12)*normrnd(0,1,[1 length(theta)]);
    noiseAmp = noise/sqrt(timedelta); %amplitude of noise
    theta = theta + timedelta * (res + noiseAmp * (rand(1,numberOfPoints) - 0.5)); %updation of theta, i.e. direction
	y(k+1,:) = [posX,posY,theta]; %Euler stepping in the solution matrix
    
%     if(mod(k,200) == 0) 
%         k %display status of execution
%     end
end

% plotting the order parameter value against step number
plot(linspace(0,Nsteps,Nsteps),orderN);
axis([0,Nsteps,0,1]);
xlabel('Time step');ylabel('Order Parameter');
% drawnow

% % write data to ./dump.txt
% timeSteps = 1:Nsteps;
% fileID = fopen('data/dump.txt','w');
% fprintf(fileID,'%d \t %6.5f \n',[timeSteps;orderN]);
% fclose(fileID);

% %movie
% for k=1:size(y)
% 	posX=y(k,1:numberOfPoints); % x position matrix
% 	posY=y(k,numberOfPoints+1:2*numberOfPoints); % y position matrix
% 	plot(posX,posY,'b.'); %plot instantaneous position
%     axis([0,L,0,L]);
%     axis square;
%     grid on;
%     %frames(k) = getframe(gcf);
%  	pause(.0005);
% end

File: szabo_grid.cpp
#include <random>
#include <iostream>
#include <fstream>
#include <cmath>
#include <stdlib.h>
#define pi 3.14159
#define N_steps 4000
#define N_particles 49
using namespace std;

/*
----------------------RUN CODE WITH COMMAND LINE INPUTS density AND noise IN THAT ORDER-------------------
was producing NaN's (for order parameter) after 800 steps or so in previous code
*/
int main(int argc, char const *argv[]) {
  double noise, rho;
	if(argc==1)
    noise = 0.6, rho = 0.3; //densty, noise
  else if(argc!=3){
    cout<<"Invalid arguments! Must enter both noise and density parameter"<<endl;
    return 0;
  }
  else{
    rho = atof(argv[1])
    noise = atof(argv[2])
  }

	double pi = 4*atan(1.0);

  double N=N_particles;
  double mu = 1, tau = 1;
  double Req = 5.0/6, R0 = 1; //radius parameters
  double v0 = 1; //self-propelling velocity

  double Fadh = 0.75, Frep = 30; //force parameters

  double L = Req * sqrt(pi*((double)(N_particles))/rho); //length of a side
  double delX = L/sqrt(N);
  double delY = delX;

  random_device rd;  //Will be used to obtain a seed for the random number engine
  mt19937 gen(rd()); //Standard mersenne_twister_engine seeded with rd()
  uniform_real_distribution<> dis(0,1); //uniform distribution on [0,1)

  double* xpos = new double[N_particles];
  double* ypos = new double[N_particles];
  double* theta = new double[N_particles];
	cout<<"l="<<L<<endl;
	cout<<"delX="<<delX<<endl;
//---------------------------initializing sol matrices-----------------------------
	int ctr = 0;
	int M = (int)sqrt(N_particles);
	cout<<"M="<<M<<endl;
  for (int i=0;i<N_particles;i++){
    xpos[i] = delX/2 + delX*ctr;
		cout<<i<<"\t"<<xpos[i]<<endl;
    ctr = (ctr+1) % M;
  }
  ctr = -1;
  for (int i=0;i<N_particles;i++){
  	if (i%M==0)
    	ctr++;
    ypos[i] = delY/2 + delY*ctr;
  }
  for (int i=0;i<N_particles;i++){
  	theta[i] = 2 * pi * dis(gen) - pi;
  }

  double delT = 0.005*R0/v0;
  double noiseAmp = noise/sqrt(delT);

  double** solX = new double*[N_steps+1];
  for(int i=0;i<N_steps+1;i++){
    solX[i] = new double[N_particles];
  }
  double** solY = new double*[N_steps+1];
  for(int i=0;i<N_steps+1;i++){
    solY[i] = new double[N_particles];
  }
  double** solTheta = new double*[N_steps+1];
  for(int i=0;i<N_steps+1;i++){
    solTheta[i] = new double[N_particles];
  }

  double* order = new double[N_steps];//order parameter of the kth step
  for (int i = 0;i<N_particles;i++){
    //cout<<i<<endl;
    solX[0][i] = xpos[i];//Initialize 0th step
    solY[0][i] = ypos[i];
    solTheta[0][i] = theta[i];
  }

	ofstream dump_pos;
	dump_pos.open("movie_dump.txt");

	ctr = 0;
  double* rhsX = new double[N_particles];
  double* rhsY = new double[N_particles];
  double* rhsTheta = new double[N_particles];
  //---------------------------main solver loop-----------------------------------
  for(int k = 0;k<N_steps;k++){
    for (int i = 0;i<N_particles;i++){
      double interaxn[2] = {0,0}; //interaction force vector for ith particle
      for (int j = 0;j<N_particles;j++){
        if (i==j)
          continue;
        double dx = solX[k][i] - solX[k][j];//dij vector = (dx,dy)
        double dy = solY[k][i] - solY[k][j];
        if (abs(dx) > 0.5*L)
          dx = dx - copysign(L,dx);
        if (abs(dy) > 0.5*L)
          dy = dy - copysign(L,dy);
        double dijSq = dx*dx + dy*dy;
        double dij = sqrt(dijSq);
        double eij[2] = {dx/dij,dy/dij};
        if (dijSq<Req*Req){
            interaxn[0] -= Frep*(dij-Req)/Req * eij[0];
            interaxn[1] -= Frep*(dij-Req)/Req * eij[1];
          }
        else if (dijSq>=Req*Req && dijSq<R0*R0){
            interaxn[0] -= Fadh*(dij-Req)/(R0-Req) * eij[0];
            interaxn[1] -= Fadh*(dij-Req)/(R0-Req) * eij[1];
          }
      }
      interaxn[0] *= mu;
      interaxn[1] *= mu;
      rhsX[i] = v0*cos(solTheta[k][i])+interaxn[0];
      rhsY[i] = v0*sin(solTheta[k][i])+interaxn[1];
    }
    double avgvel[2] = {0,0};
    for (int i = 0;i<N_particles;i++){
      double ni[2] = {cos(solTheta[k][i]),sin(solTheta[k][i])};
      double vi[2] = {rhsX[i],rhsY[i]};
      double normvi = sqrt(pow(vi[0],2)+pow(vi[1],2));
      vi[0] = vi[0]/normvi;
      vi[1] = vi[1]/normvi;
      avgvel[0] += vi[0];
      avgvel[1] += vi[1];
      rhsTheta[i] = asin(ni[0]*vi[1]-ni[1]*vi[0]);
      solX[k+1][i] = solX[k][i] + delT*rhsX[i];
      solX[k+1][i] -= L*floor(solX[k+1][i]/L);//wrapping around excesses
      solY[k+1][i] = solY[k][i] + delT*rhsY[i];
      solY[k+1][i] -= L*floor(solY[k+1][i]/L);//wrapping around excesses
      solTheta[k+1][i] = solTheta[k][i] + delT*(rhsTheta[i]/tau + noiseAmp * (dis(gen)-0.5));
			if(k%10==0){
				if(i==0)
					ctr++;
				dump_pos<<ctr<<"\t"<<i<<"\t"<<solX[k][i]<<"\t"<<solY[k][i]<<"\t"<<cos(solTheta[k][i])<<"\t"<<sin(solTheta[k][i])<<endl;
			}
    }
    order[k] = sqrt(pow(avgvel[0],2)+pow(avgvel[1],2))/N_particles;

  }
	dump_pos.close();

  ofstream dump_data;
<<<<<<< HEAD
  dump_data.open("order_dump.txt");
=======
  dump_data.open("cpp_dump.txt");
  //dump_data<<"Timestep\tOrder Parameter\n";
>>>>>>> 22e1a0e968ab087115dfc8e833aafb92f8d4acee
  for (int i=0;i<N_steps;i++){
    dump_data<<(i+1)<<"\t"<<order[i]<<"\n";
  }
  dump_data.close();
  return 0;
}
